File: kora-java-crud-submodule/kora-java-crud-submodule-app/src/main/java/ru/tinkoff/kora/example/submodule/app/controller/PetDelegate.java
package ru.tinkoff.kora.example.submodule.app.controller;

import static ru.tinkoff.kora.example.submodule.openapi.http.server.api.PetApiResponses.*;

import jakarta.annotation.Nullable;
import ru.tinkoff.kora.common.Component;
import ru.tinkoff.kora.example.submodule.app.model.PetMapper;
import ru.tinkoff.kora.example.submodule.openapi.http.server.api.PetApiDelegate;
import ru.tinkoff.kora.example.submodule.openapi.http.server.model.MessageTO;
import ru.tinkoff.kora.example.submodule.openapi.http.server.model.PetCreateTO;
import ru.tinkoff.kora.example.submodule.openapi.http.server.model.PetUpdateTO;
import ru.tinkoff.kora.example.submodule.pet.model.dao.Pet;
import ru.tinkoff.kora.example.submodule.pet.service.PetService;

@Component
public final class PetDelegate implements PetApiDelegate {

    private final PetMapper petMapper;
    private final PetService petService;

    public PetDelegate(PetMapper petMapper, PetService petService) {
        this.petMapper = petMapper;
        this.petService = petService;
    }

    @Override
    public GetPetByIdApiResponse getPetById(long petId) {
        if (petId < 0) {
            return new GetPetByIdApiResponse.GetPetById400ApiResponse(malformedId(petId));
        }

        var pet = petService.findByID(petId);
        if (pet.isPresent()) {
            var body = petMapper.asDTO(pet.get());
            return new GetPetByIdApiResponse.GetPetById200ApiResponse(body);
        } else {
            return new GetPetByIdApiResponse.GetPetById404ApiResponse(notFound(petId));
        }
    }

    @Override
    public AddPetApiResponse addPet(PetCreateTO petCreateTO) {
        var pet = petService.add(petCreateTO.name(), petCreateTO.category().name());
        var body = petMapper.asDTO(pet);
        return new AddPetApiResponse.AddPet200ApiResponse(body);
    }

    @Override
    public UpdatePetApiResponse updatePet(long petId, PetUpdateTO petUpdateTO) {
        if (petId < 0) {
            return new UpdatePetApiResponse.UpdatePet400ApiResponse(malformedId(petId));
        }

        String petCategory = petUpdateTO.category() == null
                ? null
                : petUpdateTO.category().name();
        var updated = petService.update(petId, petUpdateTO.name(), petCategory, toStatus(petUpdateTO.status()));
        if (updated.isPresent()) {
            var body = petMapper.asDTO(updated.get());
            return new UpdatePetApiResponse.UpdatePet200ApiResponse(body);
        } else {
            return new UpdatePetApiResponse.UpdatePet404ApiResponse(notFound(petId));
        }
    }

    @Override
    public DeletePetApiResponse deletePet(long petId) {
        if (petId < 0) {
            return new DeletePetApiResponse.DeletePet400ApiResponse(malformedId(petId));
        }

        if (petService.delete(petId)) {
            return new DeletePetApiResponse.DeletePet200ApiResponse(new MessageTO("Successfully deleted Pet with ID: " + petId));
        } else {
            return new DeletePetApiResponse.DeletePet404ApiResponse(notFound(petId));
        }
    }

    @Nullable
    private static Pet.Status toStatus(@Nullable PetUpdateTO.StatusEnum statusEnum) {
        if (statusEnum == null) {
            return null;
        }

        return switch (statusEnum) {
            case AVAILABLE -> Pet.Status.AVAILABLE;
            case PENDING -> Pet.Status.PENDING;
            case SOLD -> Pet.Status.SOLD;
        };
    }

    private static MessageTO notFound(long petId) {
        return new MessageTO("Pet not found for ID: " + petId);
    }

    private static MessageTO malformedId(long petId) {
        return new MessageTO("Pet malformed ID: " + petId);
    }
}

File: examples/kotlin/kora-kotlin-crud-submodule/kora-kotlin-crud-submodule-app/src/main/kotlin/ru/tinkoff/kora/kotlin/example/submodule/app/controller/PetDelegate.kt
package ru.tinkoff.kora.kotlin.example.submodule.app.controller

import ru.tinkoff.kora.common.Component
import ru.tinkoff.kora.kotlin.example.submodule.app.model.PetMapper
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.api.PetApiDelegate
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.api.PetApiResponses
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.model.MessageTO
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.model.PetCreateTO
import ru.tinkoff.kora.kotlin.example.submodule.openapi.http.server.model.PetUpdateTO
import ru.tinkoff.kora.kotlin.example.submodule.pet.model.dao.Pet
import ru.tinkoff.kora.kotlin.example.submodule.pet.service.PetService

@Component
class PetDelegate(
    private val petMapper: PetMapper,
    private val petService: PetService,
) : PetApiDelegate {
    override fun getPetById(id: Long): PetApiResponses.GetPetByIdApiResponse {
        if (id < 0) {
            return PetApiResponses.GetPetByIdApiResponse.GetPetById400ApiResponse(malformedId(id))
        }

        val pet = petService.findByID(id)
        return if (pet != null) {
            PetApiResponses.GetPetByIdApiResponse.GetPetById200ApiResponse(petMapper.asDTO(pet))
        } else {
            PetApiResponses.GetPetByIdApiResponse.GetPetById404ApiResponse(notFound(id))
        }
    }

    override fun addPet(petCreateTO: PetCreateTO): PetApiResponses.AddPetApiResponse {
        val pet = petService.add(petCreateTO.name, petCreateTO.category.name)
        return PetApiResponses.AddPetApiResponse.AddPet200ApiResponse(petMapper.asDTO(pet))
    }

    override fun updatePet(id: Long, petUpdateTO: PetUpdateTO): PetApiResponses.UpdatePetApiResponse {
        if (id < 0) {
            return PetApiResponses.UpdatePetApiResponse.UpdatePet400ApiResponse(malformedId(id))
        }

        val updated = petService.update(
            id,
            petUpdateTO.name,
            petUpdateTO.category?.name,
            toStatus(petUpdateTO.status),
        )
        return if (updated != null) {
            PetApiResponses.UpdatePetApiResponse.UpdatePet200ApiResponse(petMapper.asDTO(updated))
        } else {
            PetApiResponses.UpdatePetApiResponse.UpdatePet404ApiResponse(notFound(id))
        }
    }

    override fun deletePet(id: Long): PetApiResponses.DeletePetApiResponse {
        if (id < 0) {
            return PetApiResponses.DeletePetApiResponse.DeletePet400ApiResponse(malformedId(id))
        }

        return if (petService.delete(id)) {
            PetApiResponses.DeletePetApiResponse.DeletePet200ApiResponse(MessageTO("Successfully deleted Pet with ID: $id"))
        } else {
            PetApiResponses.DeletePetApiResponse.DeletePet404ApiResponse(notFound(id))
        }
    }

    private fun toStatus(statusEnum: PetUpdateTO.StatusEnum?): Pet.Status? {
        return when (statusEnum) {
            null -> null
            PetUpdateTO.StatusEnum.AVAILABLE -> Pet.Status.AVAILABLE
            PetUpdateTO.StatusEnum.PENDING -> Pet.Status.PENDING
            PetUpdateTO.StatusEnum.SOLD -> Pet.Status.SOLD
        }
    }

    private fun notFound(petId: Long): MessageTO = MessageTO("Pet not found for ID: $petId")
    private fun malformedId(petId: Long): MessageTO = MessageTO("Pet malformed ID: $petId")
}
